pep-0646.rst

```diff
@@ -733,28 +733,159 @@ hopefully enable a thriving ecosystem of tools for analysing and verifying
+star-unpacking in other contexts, ``_UnpackedTypeVarTuple`` would always be received
+by ``__class_getitem__`` as part of a tuple:
+
+::
+
+    class C:
+        def __class_getitem__(self, x):
+            ...
+            
+    [*(1,)]      # = [1]
+    [*(1,), 2]   # = [1, 2]
+    C[int]       # __class_getitem__ receives x=int
+    # But perhaps surprisingly:
+    C[*Ts]       # __class_getitem__ receives x=(_UnpackedTypeVarTuple, )
+    # This is consistent with:
+    C[*Ts, int]  # __class_getitem__ receives x=(_UnpackedTypeVarTuple, int)
+
+
+Second, the change would also enable star-unpacking of other structures within
+indexing operations:
+
+::
+
+    idxs_to_select = (1, 2, 3)
+    array[0, *idxs_to_select, -1]
+
+Third, more than one instance of a star-unpack can occur within an index:
+
+::
+
+    array[*idxs_to_select, *idxs_to_select]
+
+The requirement that only one unpacked TypeVarTuple may be used per list of type parameters
+would therefore be implemented in e.g. ``typing.py`` rather than at the syntax level.
+
+Finally, note that star expressions within slices remain invalid:
+
+::
 
-In both cases, at runtime the star operator would call ``Ts.__iter__()``.
-This would, in turn, return an instance of a helper class, e.g.
-``UnpackedTypeVarTuple``, whose ``repr`` would be ``*Ts``.
+    # Syntax error
+    array[*idxs_start:*idxs_end]
 
-If these grammar changes are considered too burdensome, we could instead
-simply use ``Unpack`` - though in this case it might be better for us to
-first decide whether there's a better option.
+``*args`` as a TypeVarTuple
+---------------------------
+
+The second change enables use of ``*args: *Ts``.
+
+Before:
+
+::
+
+    star_etc[StarEtc*]:
+    | '*' a=param_no_default b=param_maybe_default* c=[kwds] {
+        _PyPegen_star_etc(p, a, b, c) }
+    | '*' ',' b=param_maybe_default+ c=[kwds] {
+        _PyPegen_star_etc(p, NULL, b, c) }
+    | a=kwds { _PyPegen_star_etc(p, NULL, NULL, a) }
+
+After:
+
+::
+
+    star_etc[StarEtc*]:
+    | '*' a=param_no_default b=param_maybe_default* c=[kwds] {
+        _PyPegen_star_etc(p, a, b, c) }
+    | '*' a=param_no_default_star_annotation b=param_maybe_default* c=[kwds] {
+        _PyPegen_star_etc(p, a, b, c) }
+    | '*' ',' b=param_maybe_default+ c=[kwds] {
+        _PyPegen_star_etc(p, NULL, b, c) }
+    | a=kwds { _PyPegen_star_etc(p, NULL, NULL, a) }
+
+Where:
+
+::
+
+    param_no_default_star_annotation[arg_ty]:
+    | a=param_star_annotation ',' tc=TYPE_COMMENT? { _PyPegen_add_type_comment_to_arg(p, a, tc) }
+    | a=param_star_annotation tc=TYPE_COMMENT? &')' { _PyPegen_add_type_comment_to_arg(p, a, tc) }
+
+    param_star_annotation[arg_ty]: a=NAME b=star_annotation { _PyAST_arg(a->v.Name.id, b, NULL, EXTRA) }
+
+    star_annotation[expr_ty]: ':' a=star_expression { _PyAST_Tuple(_PyPegen_singleton_seq(p, a), Load, EXTRA) }
+
+This accomplishes the desired outcome:
+
+::
+
+    >>> def foo(*args: *b): pass
```

Comment:
So at runtime this is equivalent to `def foo(*args: tuple(b)): pass`? I think it would be good to say so explicitly.

Reply:
Not quite, but I can see how this is confusing. I've added some extra detail - does this clarify?

Reply:
Thanks, but I think this still doesn't tell me what happens in terms of Python code execution. I would like to see an equivalent in terms of existing syntax; if it's not `tuple(bs)` what is it?

Reply:
Ohhh, sorry - for some reason I interpreter `tuple(b)` as `(b,)`. Yes, `tuple(b)` is exactly right. Is this clear in the PEP now? (Sorry for the commit spam; a forced push didn't behave as I expected...)

Reply:
Great. Resolving.